stages/org.osbuild.coreos.live-artifacts.mono

```diff
@@ -695,24 +753,10 @@ def main(workdir, tree, inputs, options, loop_client):
     # version which may impede uniqueness.
     volid = name_version[0:32]
 
-    tmpisofile = os.path.join(workdir, 'live.iso')
-    tmpisoroot = os.path.join(workdir, 'iso')
-    tmpisocoreos = os.path.join(tmpisoroot, 'coreos')
-    tmpisoimages = os.path.join(tmpisoroot, 'images')
-    tmpisoimagespxe = os.path.join(tmpisoimages, 'pxeboot')
-    tmpisoisolinux = os.path.join(tmpisoroot, 'isolinux')
-    # contents of initramfs on both PXE and ISO
-    tmpinitrd_base = os.path.join(workdir, 'initrd')
-    # contents of rootfs image
-    tmpinitrd_rootfs = os.path.join(workdir, 'initrd-rootfs')
-
-    for d in (tmpisoroot, tmpisocoreos, tmpisoimages, tmpisoimagespxe,
-              tmpisoisolinux, tmpinitrd_base, tmpinitrd_rootfs):
-        os.mkdir(d)
```

Comment:
> What if instead we defined a dictionary with all of these values in it (the dictionary can be passed a signle argument to functions, thus solving that problem we had)

This thought crossed my mind as well.  I arbitrarily chose the other option of doing it with functions.  Since you suggested this, let's do it this way.

Reply:
One downside with this new approach though is that it's harder to catch typos or mistakes with a linter.  `paths["isp"]["coreos"]` wont get caught by static analysis, whereas `mkpath_tmpispcoreos(workdir)` will.

Reply:
I think I'm OK with that.. we do have a test that does an actual functional run of the code that would catch it.

Reply:
Here's a diff of just this change, from functions to the dictionary.  I kept the functions approach in a separate branch to compare functionality in case I mixed up some paths.
```patch
diff --git a/stages/org.osbuild.coreos.live-artifacts.mono b/stages/org.osbuild.coreos.live-artifacts.mono
index 65626dbe..05bfb51b 100755
--- a/stages/org.osbuild.coreos.live-artifacts.mono
+++ b/stages/org.osbuild.coreos.live-artifacts.mono
@@ -182,7 +182,7 @@ def parse_metal_inputs(inputs):
     return metal_file, metal4k_file
 
 
-def genisoargs_x86_64(paths):
+def genisoargs_x86_64(workdir):
     # Install binaries from syslinux package
     isolinuxfiles = [('/usr/share/syslinux/isolinux.bin', 0o755),
                      ('/usr/share/syslinux/ldlinux.c32', 0o755),
@@ -190,7 +190,7 @@ def genisoargs_x86_64(paths):
                      ('/usr/share/syslinux/libutil.c32', 0o755),
                      ('/usr/share/syslinux/vesamenu.c32', 0o755)]
     for src, mode in isolinuxfiles:
-        dst = os.path.join(paths["iso"]["isolinux"], os.path.basename(src))
+        dst = os.path.join(mkpath_tmpisoisolinux(workdir), os.path.basename(src))
         shutil.copyfile(src, dst)
         os.chmod(dst, mode)
 
@@ -202,8 +202,8 @@ def genisoargs_x86_64(paths):
                 '-boot-info-table']
 
 
-def genisoargs_ppc64le(paths, kargs_json, volid):
-    tmpisoroot = paths["iso"]["root"]
+def genisoargs_ppc64le(workdir, kargs_json, volid):
+    tmpisoroot = mkpath_tmpisoroot(workdir)
     os.makedirs(os.path.join(tmpisoroot, 'boot/grub'))
     # can be EFI/fedora or EFI/redhat
     grubpath = ensure_glob(os.path.join(tmpisoroot, 'EFI/*/grub.cfg'))
@@ -222,10 +222,9 @@ def genisoargs_ppc64le(paths, kargs_json, volid):
     return ['grub2-mkrescue', '-volid', volid]
 
 
-def genisoargs_s390x(paths, kernel_img, test_fixture, volid, name_version):
-    tmpisoimages = paths["iso"]["images"]
-    tmpisoroot = paths["iso"]["root"]
-    pxeimages = paths["iso"]["pxeboot"]
+def genisoargs_s390x(workdir, kernel_img, test_fixture, volid, name_version):
+    tmpisoimages = mkpath_tmpisoimages(workdir)
+    tmpisoroot = mkpath_tmpisoroot(workdir)
     # Reserve 32MB for the kernel, starting memory address of the initramfs
     # See https://github.com/weldr/lorax/blob/master/share/templates.d/99-generic/s390.tmpl
     INITRD_ADDRESS = '0x02000000'
@@ -242,10 +241,10 @@ def genisoargs_s390x(paths, kernel_img, test_fixture, volid, name_version):
     # TODO: check if the kernel.img move can be extracted
     # s390x's z/VM CMS files are limited to 8 char for filenames and extensions
     # Also it is nice to keep naming convetion with Fedora/RHEL for existing users and code
-    kernel_dest = os.path.join(pxeimages, 'kernel.img')
-    shutil.move(os.path.join(pxeimages, kernel_img), kernel_dest)
+    kernel_dest = os.path.join(mkpath_tmpisoimagespxe(workdir), 'kernel.img')
+    shutil.move(os.path.join(mkpath_tmpisoimagespxe(workdir), kernel_img), kernel_dest)
 
-    iso_initramfs = paths["iso"]["initramfs"]
+    iso_initramfs = mkpath_iso_initramfs(workdir)
     if test_fixture:
         # truncate it to 128k so it includes the offsets to the initrd and kargs
         # https://github.com/ibm-s390-linux/s390-tools/blob/032304d5034e/netboot/mk-s390image#L21-L24
@@ -289,7 +288,7 @@ def genisoargs_s390x(paths, kernel_img, test_fixture, volid, name_version):
             os.path.join(os.path.relpath(tmpisoimages, tmpisoroot), 'cdboot.img')]
 
 
-def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img, iso_rootfs, version, blsentry_kargs):
+def genisoimage(workdir, tree, filenames, deployed_tree, loop_client, kernel_img, iso_rootfs, version, blsentry_kargs):
     """
     Generate the ISO image.
     Based on the Lorax templates [1].
@@ -321,7 +320,7 @@ def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img,
     test_fixture = check_for_test_fixture(deployed_tree)
     basearch = os.uname().machine
 
-    kargs_json = init_iso_kargs(deployed_tree, paths, blsentry_kargs, volid)
+    kargs_json = init_iso_kargs(deployed_tree, workdir, blsentry_kargs, volid)
 
     # Add placeholder for Ignition CPIO file.  This allows an external tool,
     # `coreos-installer iso ignition embed`, to modify an existing ISO image
@@ -334,18 +333,18 @@ def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img,
     # handled lower down
     igninfo_json = {}
     if basearch != 's390x':
-        with open(os.path.join(paths["iso"]["images"], 'ignition.img'), 'wb') as fdst:
+        with open(os.path.join(mkpath_tmpisoimages(workdir), 'ignition.img'), 'wb') as fdst:
             fdst.write(bytes(IGNITION_IMG_SIZE))
         igninfo_json = {'file': 'images/ignition.img'}
 
     # For x86_64 legacy boot (BIOS) booting
     if basearch == "x86_64":
-        genisoargs += genisoargs_x86_64(paths)
+        genisoargs += genisoargs_x86_64(workdir)
     elif basearch == "ppc64le":
-        genisoargs = genisoargs_ppc64le(paths, kargs_json, volid)
+        genisoargs = genisoargs_ppc64le(workdir, kargs_json, volid)
     elif basearch == "s390x":
-        genisoargs = genisoargs_s390x(paths, kernel_img, test_fixture, volid, name_version)
-        igninfo_json = update_image_offsets(paths, kargs_json)
+        genisoargs = genisoargs_s390x(workdir, kernel_img, test_fixture, volid, name_version)
+        igninfo_json = update_image_offsets(workdir, kargs_json)
         # This name change happened in the code extracted into genisoargs_s390x(). Now it's part of the function, so it
         # doesn't leak into the main() scope and we need to set it here again. Splitting the genisoargs_s390x() function
         # further, making the file rename part its own function if possible, will fix this issue.
@@ -353,52 +352,52 @@ def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img,
 
     # Drop zipl.prm as it was either unused (!s390x) or is now no longer
     # needed (s390x)
-    os.unlink(os.path.join(paths["iso"]["root"], 'zipl.prm'))
+    os.unlink(os.path.join(mkpath_tmpisoroot(workdir), 'zipl.prm'))
 
     # For x86_64 and aarch64 UEFI booting
     if basearch in ("x86_64", "aarch64"):
-        efiboot_path = mkefiboot(paths, deployed_tree, kargs_json, volid, loop_client)
+        efiboot_path = mkefiboot(workdir, deployed_tree, kargs_json, volid, loop_client)
         genisoargs += ['-eltorito-alt-boot', '-efi-boot', efiboot_path, '-no-emul-boot']
 
     # We've done everything that might affect kargs, so filter out any files
     # that no longer exist and write out the kargs JSON if it lists any files
     kargs_json['files'] = [f for f in kargs_json['files']
-                           if os.path.exists(os.path.join(paths["iso"]["root"], f['path']))]
+                           if os.path.exists(os.path.join(mkpath_tmpisoroot(workdir), f['path']))]
     kargs_json['files'].sort(key=lambda f: f['path'])
     if kargs_json['files']:
         # Store the location of "karg embed areas" for use by
         # `coreos-installer iso kargs modify`
-        with open(os.path.join(paths["iso"]["coreos"], kargs_file), 'w', encoding='utf8') as fh:
+        with open(os.path.join(mkpath_tmpisocoreos(workdir), kargs_file), 'w', encoding='utf8') as fh:
             json.dump(kargs_json, fh, indent=2, sort_keys=True)
             fh.write('\n')
 
     # Write out the igninfo.json file. This is used by coreos-installer to know
     # how to embed the Ignition config.
-    with open(os.path.join(paths["iso"]["coreos"], igninfo_file), 'w', encoding='utf8') as fh:
+    with open(os.path.join(mkpath_tmpisocoreos(workdir), igninfo_file), 'w', encoding='utf8') as fh:
         json.dump(igninfo_json, fh, indent=2, sort_keys=True)
         fh.write('\n')
 
     # Define inputs and outputs
-    genisoargs_final = genisoargs + ['-o', paths["live.iso"], paths["iso"]["root"]]
+    genisoargs_final = genisoargs + ['-o', mkpath_tmpisofile(workdir), mkpath_tmpisoroot(workdir)]
 
-    miniso_data = os.path.join(paths["iso"]["coreos"], "miniso.dat")
+    miniso_data = os.path.join(mkpath_tmpisocoreos(workdir), "miniso.dat")
     with open(miniso_data, 'wb') as f:
         f.truncate(MINISO_DATA_FILE_SIZE)
 
     if test_fixture:
-        trim_for_test_fixtures(paths)
+        trim_for_test_fixtures(workdir)
 
     subprocess.check_call(genisoargs_final)
 
     # Add MBR, and GPT with ESP, for x86_64 BIOS/UEFI boot when ISO is
     # copied to a USB stick
     if basearch == "x86_64":
-        subprocess.check_call(['/usr/bin/isohybrid', '--uefi', paths["live.iso"]])
+        subprocess.check_call(['/usr/bin/isohybrid', '--uefi', mkpath_tmpisofile(workdir)])
 
     # Copy to final output locations the things that won't change from
     # this point forward. We do it here because we unlink the the rootfs below.
-    cp_reflink(os.path.join(paths["iso"]["pxeboot"], kernel_img), output_kernel)
-    cp_reflink(paths["iso"]["initramfs"], output_initramfs)
+    cp_reflink(os.path.join(mkpath_tmpisoimagespxe(workdir), kernel_img), output_kernel)
+    cp_reflink(mkpath_iso_initramfs(workdir), output_initramfs)
     cp_reflink(iso_rootfs, output_rootfs)
 
     # Here we generate a minimal ISO purely for the purpose of
@@ -417,8 +416,8 @@ def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img,
     # coreos-installer takes care of removing it.
     os.unlink(iso_rootfs)
     os.unlink(miniso_data)
-    tmpisofile = paths["live.iso"]
-    subprocess.check_call(genisoargs + ['-o', f'{tmpisofile}.minimal', paths["iso"]["root"]])
+    tmpisofile = mkpath_tmpisofile(workdir)
+    subprocess.check_call(genisoargs + ['-o', f'{tmpisofile}.minimal', mkpath_tmpisoroot(workdir)])
     if basearch == "x86_64":
         subprocess.check_call(['/usr/bin/isohybrid', '--uefi', f'{tmpisofile}.minimal'])
     with Chroot(deployed_tree, bind_mounts=["/run"]) as chroot:
@@ -428,16 +427,15 @@ def genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img,
     return tmpisofile
 
 
-def update_image_offsets(paths, kargs_json):
+def update_image_offsets(workdir, kargs_json):
     # Get the kargs and initramfs offsets in the cdboot.img. For more info, see:
     # https://github.com/ibm-s390-linux/s390-tools/blob/032304d5034e/netboot/mk-s390image#L21-L23
     CDBOOT_IMG_OFFS_INITRD_START_BYTES = 66568
     CDBOOT_IMG_OFFS_KARGS_START_BYTES = 66688
     CDBOOT_IMG_OFFS_KARGS_MAX_SIZE = 896
 
-    iso_initramfs = paths["iso"]["initramfs"]
-    iso_images = paths["iso"]["images"]
-    with open(os.path.join(iso_images, 'cdboot.img'), 'rb') as f:
+    iso_initramfs = mkpath_iso_initramfs(workdir)
+    with open(os.path.join(mkpath_tmpisoimages(workdir), 'cdboot.img'), 'rb') as f:
         f.seek(CDBOOT_IMG_OFFS_INITRD_START_BYTES)
         offset = struct.unpack(">Q", f.read(8))[0]
 
@@ -464,7 +462,7 @@ def update_image_offsets(paths, kargs_json):
     }
 
 
-def mkefiboot(paths, deployed_tree, kargs_json, volid, loop_client):
+def mkefiboot(workdir, deployed_tree, kargs_json, volid, loop_client):
     """
     Creates an efi boot image, required for EFI booting the ISO.
 
@@ -487,18 +485,18 @@ def mkefiboot(paths, deployed_tree, kargs_json, volid, loop_client):
                 tarinfo.mode = 0o0644
             return tarinfo
 
-        efidir = paths["efi"]
-        shutil.copytree(os.path.join(deployed_tree, 'usr/lib/bootupd/updates/EFI'), efidir)
+        tmpimageefidir = os.path.join(workdir, "efi")
+        shutil.copytree(os.path.join(deployed_tree, 'usr/lib/bootupd/updates/EFI'), tmpimageefidir)
 
         # Find name of vendor directory
-        vendor_ids = [n for n in os.listdir(efidir) if n != "BOOT"]
+        vendor_ids = [n for n in os.listdir(tmpimageefidir) if n != "BOOT"]
         if len(vendor_ids) != 1:
             raise ValueError(f"did not find exactly one EFI vendor ID: {vendor_ids}")
         vendor_id = vendor_ids[0]
 
         # Always replace live/EFI/{vendor} to actual live/EFI/{vendor_id}
         # https://github.com/openshift/os/issues/954
-        dfd = os.open(paths["iso"]["root"], os.O_RDONLY)
+        dfd = os.open(mkpath_tmpisoroot(workdir), os.O_RDONLY)
         grubfilepath = ensure_glob('EFI/*/grub.cfg', dir_fd=dfd)
         if len(grubfilepath) != 1:
             raise ValueError(f'Found != 1 grub.cfg files: {grubfilepath}')
@@ -519,21 +517,21 @@ def mkefiboot(paths, deployed_tree, kargs_json, volid, loop_client):
         # A future shim release will merge fallback.efi into the main
         # shim binary and enable the fallback behavior when the CSV
         # exists.  But for now, fail if fallback.efi is missing.
-        for path in ensure_glob(os.path.join(efidir, "BOOT", "fb*.efi")):
+        for path in ensure_glob(os.path.join(tmpimageefidir, "BOOT", "fb*.efi")):
             os.unlink(path)
-        for path in ensure_glob(os.path.join(efidir, vendor_id, "BOOT*.CSV")):
+        for path in ensure_glob(os.path.join(tmpimageefidir, vendor_id, "BOOT*.CSV")):
             os.unlink(path)
 
         # Drop vendor copies of shim; we already have it in BOOT*.EFI in
         # BOOT
-        for path in ensure_glob(os.path.join(efidir, vendor_id, "shim*.efi")):
+        for path in ensure_glob(os.path.join(tmpimageefidir, vendor_id, "shim*.efi")):
             os.unlink(path)
 
         # Consolidate remaining files into BOOT.  shim needs GRUB to be
         # there, and the rest doesn't hurt.
-        for path in ensure_glob(os.path.join(efidir, vendor_id, "*")):
-            shutil.move(path, os.path.join(efidir, "BOOT"))
-        os.rmdir(os.path.join(efidir, vendor_id))
+        for path in ensure_glob(os.path.join(tmpimageefidir, vendor_id, "*")):
+            shutil.move(path, os.path.join(tmpimageefidir, "BOOT"))
+        os.rmdir(os.path.join(tmpimageefidir, vendor_id))
 
         # Inject a stub grub.cfg pointing to the one in the main ISO image.
         #
@@ -549,7 +547,7 @@ def mkefiboot(paths, deployed_tree, kargs_json, volid, loop_client):
         # Torito at all.  In that case, GRUB thinks it booted from a
         # partition of the disk (a fake ESP created by isohybrid,
         # pointing to efiboot.img) and needs a grub.cfg there.
-        with open(os.path.join(efidir, "BOOT", "grub.cfg"), "w", encoding='utf8') as fh:
+        with open(os.path.join(tmpimageefidir, "BOOT", "grub.cfg"), "w", encoding='utf8') as fh:
             fh.write(f'''search --label "{volid}" --set root --no-floppy
 set prefix=($root)/EFI/{vendor_id}
 echo "Booting via ESP..."
@@ -561,10 +559,10 @@ boot
         # Manually construct the tarball to ensure proper permissions and ownership
         efitarfile = tempfile.NamedTemporaryFile(suffix=".tar")
         with tarfile.open(efitarfile.name, "w:", dereference=True) as tar:
-            tar.add(efidir, arcname="/EFI", filter=strip)
+            tar.add(tmpimageefidir, arcname="/EFI", filter=strip)
 
         # Create the efiboot.img file in the images/ dir
-        efibootfile = os.path.join(paths["iso"]["images"], 'efiboot.img')
+        efibootfile = os.path.join(mkpath_tmpisoimages(workdir), 'efiboot.img')
         make_efi_bootfile(loop_client, input_tarball=efitarfile.name, output_efiboot_img=efibootfile)
 
     return "images/efiboot.img"
@@ -660,7 +658,7 @@ def mksquashfs_metal(workdir, img_metal, loop_client):
     return blsentry_kargs
 
 
-def init_iso_kargs(deployed_tree, paths, blsentry_kargs, volid):
+def init_iso_kargs(deployed_tree, workdir, blsentry_kargs, volid):
     # Filter kernel arguments for substituting into ISO bootloader
     kargs_array = [karg for karg in blsentry_kargs
                    if karg.split('=')[0] not in LIVE_EXCLUDE_KARGS]
@@ -675,7 +673,7 @@ def init_iso_kargs(deployed_tree, paths, blsentry_kargs, volid):
     # Grab all the contents from the live dir from the configs
     for srcdir, _, filenames in os.walk(srcdir_prefix):
         dir_suffix = srcdir.replace(srcdir_prefix, '', 1)
-        dstdir = os.path.join(paths["iso"]["root"], dir_suffix)
+        dstdir = os.path.join(mkpath_tmpisoroot(workdir), dir_suffix)
         if not os.path.exists(dstdir):
             os.mkdir(dstdir)
         for filename in filenames:
@@ -737,37 +735,37 @@ def check_for_test_fixture(deployed_tree):
     return os.path.exists(os.path.join(deployed_tree, 'usr/share/coreos-assembler/coreos-installer-test-fixture'))
 
 
-def trim_for_test_fixtures(paths):
+def trim_for_test_fixtures(workdir):
     """
     Replaces the large images (efiboot.img, rootfs.img, etc) with simple text files to aid in test fixture creation.
     """
-    images = paths["iso"]["images"]
-    isolinux = paths["iso"]["isolinux"]
-    pxeboot = paths["iso"]["pxeboot"]
-    with open(os.path.join(images, 'efiboot.img'), 'w', encoding='utf8') as fh:
+    tmpisoimages = mkpath_tmpisoimages(workdir)
+    tmpisoisolinux = mkpath_tmpisoisolinux(workdir)
+    tmpisoimagespxe = mkpath_tmpisoimagespxe(workdir)
+    with open(os.path.join(tmpisoimages, 'efiboot.img'), 'w', encoding='utf8') as fh:
         fh.write('efiboot.img\n')
-    with open(os.path.join(pxeboot, 'rootfs.img'), 'w', encoding='utf8') as fh:
+    with open(os.path.join(tmpisoimagespxe, 'rootfs.img'), 'w', encoding='utf8') as fh:
         fh.write('rootfs data\n')
-    with open(os.path.join(pxeboot, 'initrd.img'), 'w', encoding='utf8') as fh:
+    with open(os.path.join(tmpisoimagespxe, 'initrd.img'), 'w', encoding='utf8') as fh:
         fh.write('initrd data\n')
-    with open(os.path.join(pxeboot, 'vmlinuz'), 'w', encoding='utf8') as fh:
+    with open(os.path.join(tmpisoimagespxe, 'vmlinuz'), 'w', encoding='utf8') as fh:
         fh.write('the kernel\n')
     # this directory doesn't exist on s390x
-    if os.path.isdir(isolinux):
-        with open(os.path.join(isolinux, 'isolinux.bin'), 'rb+') as fh:
+    if os.path.isdir(tmpisoisolinux):
+        with open(os.path.join(tmpisoisolinux, 'isolinux.bin'), 'rb+') as fh:
             flen = fh.seek(0, 2)
             fh.truncate(0)
             fh.truncate(flen)
             fh.seek(64)
             # isohybrid checks for this magic
             fh.write(b'\xfb\xc0\x78\x70')
-        for f in ensure_glob(os.path.join(isolinux, '*.c32')):
+        for f in ensure_glob(os.path.join(tmpisoisolinux, '*.c32')):
             os.unlink(f)
-        for f in ensure_glob(os.path.join(isolinux, '*.msg')):
+        for f in ensure_glob(os.path.join(tmpisoisolinux, '*.msg')):
             os.unlink(f)
 
 
-def mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, paths, os_release):
+def mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, workdir, os_release):
     """
     Add osmet files. Use a chroot here to use the same coreos-installer as is
     in the artifacts we are building. Use the deployed_tree as the chroot since
@@ -778,7 +776,7 @@ def mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, paths, os
             img_name = os.path.basename(img)
             print(f'Generating osmet file for {img_name} image')
             img_checksum = checksum.hexdigest_file(loopdev, "sha256")
-            img_osmet_file = os.path.join(paths["initrd-rootfs"], f'{img_name}.osmet')
+            img_osmet_file = os.path.join(mkpath_tmpinitrd_rootfs(workdir), f'{img_name}.osmet')
             with Chroot(deployed_tree, bind_mounts=["/run", "/tmp"]) as chroot:
                 cmd = ['coreos-installer', 'pack', 'osmet', loopdev,
                        '--description', os_release['PRETTY_NAME'],
@@ -786,36 +784,55 @@ def mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, paths, os
                 chroot.run(cmd, check=True)
 
 
-def mk_paths(workdir):
-    """
-    Returns a dictionary with all the paths under workdir needed throughout multiple functions in the stage.
-    """
-    return {
-        "live.iso": os.path.join(workdir, "live.iso"),
-        "efi": os.path.join(workdir, "efi"),
-        "iso": {
-            "root": os.path.join(workdir, "iso"),
-            "coreos": os.path.join(workdir, "iso", "coreos"),
-            "isolinux": os.path.join(workdir, "iso", "isolinux"),
-            "images": os.path.join(workdir, "iso", "images"),
-            "pxeboot": os.path.join(workdir, "iso", "images", "pxeboot"),
-            "initramfs": os.path.join(workdir, "iso", "images", "pxeboot", "initrd.img"),
-        },
-        "initrd": os.path.join(workdir, "initrd"),
-        "initrd-rootfs": os.path.join(workdir, "initrd-rootfs"),
-    }
+def mkpath_tmpisofile(workdir):
+    return os.path.join(workdir, 'live.iso')
+
+
+def mkpath_tmpisoroot(workdir):
+    return os.path.join(workdir, 'iso')
+
+
+def mkpath_tmpisocoreos(workdir):
+    return os.path.join(mkpath_tmpisoroot(workdir), 'coreos')
+
+
+def mkpath_tmpisoimages(workdir):
+    return os.path.join(mkpath_tmpisoroot(workdir), 'images')
+
+
+def mkpath_tmpisoimagespxe(workdir):
+    return os.path.join(mkpath_tmpisoimages(workdir), 'pxeboot')
+
+
+def mkpath_tmpisoisolinux(workdir):
+    return os.path.join(mkpath_tmpisoroot(workdir), 'isolinux')
+
+
+def mkpath_tmpinitrd_base(workdir):
+    return os.path.join(workdir, 'initrd')
+
+
+def mkpath_tmpinitrd_rootfs(workdir):
+    return os.path.join(workdir, 'initrd-rootfs')
+
+
+def mkpath_iso_initramfs(workdir):
+    return os.path.join(mkpath_tmpisoimagespxe(workdir), "initrd.img")
 
 
-def mk_workdirs(paths):
-    isoroot = paths["iso"]["root"]
-    isocoreos = paths["iso"]["coreos"]
-    isoimages = paths["iso"]["images"]
-    isoimagespxe = paths["iso"]["pxeboot"]
-    isolinux = paths["iso"]["isolinux"]
-    initrd = paths["initrd"]
-    initrd_rootfs = paths["initrd-rootfs"]
+def mk_workdirs(workdir):
+    tmpisoroot = mkpath_tmpisoroot(workdir)
+    tmpisocoreos = mkpath_tmpisocoreos(workdir)
+    tmpisoimages = mkpath_tmpisoimages(workdir)
+    tmpisoimagespxe = mkpath_tmpisoimagespxe(workdir)
+    tmpisoisolinux = mkpath_tmpisoisolinux(workdir)
+    # contents of initramfs on both PXE and ISO
+    tmpinitrd_base = mkpath_tmpinitrd_base(workdir)
+    # contents of rootfs image
+    tmpinitrd_rootfs = mkpath_tmpinitrd_rootfs(workdir)
 
-    for d in (isoroot, isocoreos, isoimages, isoimagespxe, isolinux, initrd, initrd_rootfs):
+    for d in (tmpisoroot, tmpisocoreos, tmpisoimages, tmpisoimagespxe,
+              tmpisoisolinux, tmpinitrd_base, tmpinitrd_rootfs):
         os.mkdir(d)
 
 
@@ -857,8 +874,7 @@ def main(workdir, tree, inputs, options, loop_client):
     os_release = osrelease.parse_files(os.path.join(deployed_tree, 'etc', 'os-release'))
     version = os_release['OSTREE_VERSION']
 
-    paths = mk_paths(workdir)
-    mk_workdirs(paths)
+    mk_workdirs(workdir)
 
     # convention for kernel and rootfs names
     kernel_img = 'vmlinuz'
@@ -874,18 +890,19 @@ def main(workdir, tree, inputs, options, loop_client):
 
     # copy those files out of the ostree into the iso root dir
     initramfs_img = 'initramfs.img'
+    tmpisoimagespxe = mkpath_tmpisoimagespxe(workdir)
     for file in [kernel_img, initramfs_img]:
         src = os.path.join(modules_dir, moduledir, file)
-        dst = os.path.join(paths["iso"]["pxeboot"], file)
+        dst = os.path.join(tmpisoimagespxe, file)
         if file == initramfs_img:
-            dst = paths["iso"]["initramfs"]
+            dst = mkpath_iso_initramfs(workdir)
         shutil.copyfile(src, dst)
         # initramfs isn't world readable by default so let's open up perms
         os.chmod(dst, 0o644)
 
     # Generate initramfs stamp file indicating that this is a live
     # initramfs.  Store the build ID in it.
-    stamppath = os.path.join(paths["initrd"], 'etc/coreos-live-initramfs')
+    stamppath = os.path.join(mkpath_tmpinitrd_base(workdir), 'etc/coreos-live-initramfs')
     os.makedirs(os.path.dirname(stamppath), exist_ok=True)
     with open(stamppath, 'w', encoding='utf8') as fh:
         fh.write(version + '\n')
@@ -893,7 +910,7 @@ def main(workdir, tree, inputs, options, loop_client):
     # Generate rootfs stamp file with the build ID, indicating that the
     # rootfs has been appended and confirming that initramfs and rootfs are
     # from the same build.
-    stamppath = os.path.join(paths["initrd-rootfs"], 'etc/coreos-live-rootfs')
+    stamppath = os.path.join(mkpath_tmpinitrd_rootfs(workdir), 'etc/coreos-live-rootfs')
     os.makedirs(os.path.dirname(stamppath), exist_ok=True)
     with open(stamppath, 'w', encoding='utf8') as fh:
         fh.write(version + '\n')
@@ -902,22 +919,22 @@ def main(workdir, tree, inputs, options, loop_client):
     # coreos-installer {iso|pxe} customize.  Put it in the initramfs for
     # pxe customize and the ISO for iso customize.
     features = json.dumps(get_os_features(tree=deployed_tree), indent=2, sort_keys=True) + '\n'
-    featurespath = os.path.join(paths["initrd"], 'etc/coreos/features.json')
+    featurespath = os.path.join(mkpath_tmpinitrd_base(workdir), 'etc/coreos/features.json')
     os.makedirs(os.path.dirname(featurespath), exist_ok=True)
     with open(featurespath, 'w', encoding='utf8') as fh:
         fh.write(features)
-    with open(os.path.join(paths["iso"]["coreos"], 'features.json'), 'w', encoding='utf8') as fh:
+    with open(os.path.join(mkpath_tmpisocoreos(workdir), 'features.json'), 'w', encoding='utf8') as fh:
         fh.write(features)
 
-    mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, paths, os_release)
+    mk_osmet_files(deployed_tree, img_metal, img_metal4k, loop_client, workdir, os_release)
 
     blsentry_kargs = mksquashfs_metal(workdir, img_metal, loop_client)
 
     # Generate rootfs image
-    iso_rootfs = os.path.join(paths["iso"]["pxeboot"], rootfs_img)
+    iso_rootfs = os.path.join(tmpisoimagespxe, rootfs_img)
     # The rootfs must be uncompressed because the ISO mounts root.squashfs
     # directly from the middle of the file
-    extend_initramfs(initramfs=iso_rootfs, tree=paths["initrd-rootfs"], compress=False)
+    extend_initramfs(initramfs=iso_rootfs, tree=mkpath_tmpinitrd_rootfs(workdir), compress=False)
     # Check that the root.squashfs magic number is in the offset hardcoded
     # in sysroot.mount in 20live/live-generator
     with open(iso_rootfs, 'rb') as fh:
@@ -925,13 +942,14 @@ def main(workdir, tree, inputs, options, loop_client):
         if fh.read(4) != b'hsqs':
             raise ValueError("root.squashfs not at expected offset in rootfs image")
     # Save stream hash of rootfs for verifying out-of-band fetches
-    os.makedirs(os.path.join(paths["initrd"], 'etc'), exist_ok=True)
-    make_stream_hash(iso_rootfs, os.path.join(paths["initrd"], 'etc/coreos-live-want-rootfs'))
+    tmpinitrd_base = mkpath_tmpinitrd_base(workdir)
+    os.makedirs(os.path.join(tmpinitrd_base, 'etc'), exist_ok=True)
+    make_stream_hash(iso_rootfs, os.path.join(tmpinitrd_base, 'etc/coreos-live-want-rootfs'))
     # Add common content
-    extend_initramfs(initramfs=paths["iso"]["initramfs"], tree=paths["initrd"])
+    extend_initramfs(initramfs=mkpath_iso_initramfs(workdir), tree=tmpinitrd_base)
 
     filenames = options['filenames']
-    tmpisofile = genisoimage(paths, tree, filenames, deployed_tree, loop_client, kernel_img, iso_rootfs, version,
+    tmpisofile = genisoimage(workdir, tree, filenames, deployed_tree, loop_client, kernel_img, iso_rootfs, version,
                              blsentry_kargs)
     # Copy the ISO to the final output location
     output_iso = os.path.join(tree, filenames['live-iso'])
```